src/analyzing/AST_to_IL.ml

```diff
@@ -178,7 +187,27 @@ let add_instr env instr = Common.push (mk_s (Instr instr)) env.stmts
  * itself is already a variable! *)
 let mk_aux_var ?str env tok exp =
   match exp.e with
-  | Fetch ({ base = Var var; rev_offset = []; _ } as lval) -> (var, lval)
+  | Fetch ({ base = Var var; rev_offset = []; _ } as lval) ->
```

Comment:
This seems like a dangerous place to put this code. mk_aux_var is used in many context.
Wouldn't it be possible instead to add some special code in stmt_aux() for the ExprStmt case
and detect when no statement was generated and create a fake assign there? 
What do you think @IagoAbal ?

Reply:
Then we would no need treat_expr_as_stmt guard; each time we have an ExprStmt that does not generate
any instruction, that probably means something is wrong and we should still generate a node
(in theory we could issue a warning because in your Rust example the first two x are wrong; people should not write such code).

Reply:
Also agree with Yoann here, we could just unconditionally add this assignment in `stmt_aux` for the `ExprStmt` case. We could still make use of `treat_expr_as_stmt` to determine whether the value of `~void` (right now we always set it to `true` in an `ExprStmt`).

Reply:
Sounds good, thanks! I'll come back with an update.

Reply:
Some updates:

At first, I thought it wouldn't hurt to unconditionally create a fake assignments for the `ExprStmt` case in `stmt_aux` and not call `mk_aux_var`, but there are some issues with this approach.

The code

```
x = 0
```

results in the IL
```
x = 0
tmp = x
```

where both `x`s point to the same generic ast.

During constant propagation, because the second `x` is [on the RHS, its svalue is updated](https://github.com/returntocorp/semgrep/blob/2691ed6fe4feb6499438358532b6929c5b37f306/src/analyzing/Dataflow_svalue.ml#L654-L655) to 0.

This means 0 matches both with x and 0 in `x=0`, but we only want to match with the rhs for constant propagation so this turned out to be incorrect.

Checking for empty results like @aryx suggested seems to be a good approach. I implemented it and it works. I'll send the cleaned up code next week. I'm reluctant to add warnings because while the first two `x`s in
```
  fn f(x:i32) {
    x;
    x;
    x;
  }
```
doesn't make sense, the final `x` is a legitimate one.

I couldn't find good use of setting `~void` here. The `~void` parameter doesn't seem to have any affect on the AST_Generic.N case that I'm interested here. I may be missing what you're trying to suggest.

Reply:
Updated the code.